architecture-docs/adr-001.md

```diff
@@ -0,0 +1,85 @@
+validator tracking, etc.
+* tx-validation enclave (TVE): sealed transaction data (of valid obfuscated transactions that have outputs)
+
+The reason for having two processes is that SGX SDK compilation is different and needs Intel SGX SDK tooling,
+so for the development convenience, the transaction validation code that needs to execute in an enclave
+is isolated. The inter-process communication is over a simple REQ-REP 0MQ socket.
+
+*Problem 1: These two storage locations need to be "in sync"*:
+
+when an obfuscated transaction arrives that spends some transaction outputs, chain-abci will do a basic validation and check if they are unspent and forward it to TVE (assuming its storage contains
+sealed transaction data of respective outputs). There is currently a naive check that TVE
+stores the latest app hash provided by chain-abci; and upon a startup, chain-abci cross-checks if TVE is in sync with it. This leads to various errors and annoyances that are usually resolved by removing all storage and syncing from scratch.
+
+*Problem 2: Transaction querying*:
+
+As wallet / client-* may be lightweight client and not have access to TEE directly, it will connect to one remotely.
+For this purpose, there is transaction query enclave (TQE). See [this document](https://github.com/crypto-com/chain-docs/blob/master/plan.md#transaction-query-enclave-tqe-optional--for-client-infrastructure) for more details.
+
+There are two flows (over an attested secure channel):
+
+1. retrieving transactions: client submits transaction identifiers signed by its view key, and TQE replies with matching transaction data. For this workflow, TQE contacts TVE over REQ-REP 0MQ socket to retrieve data.
+
+2. submitting new transactions: client submits a new transaction, TQE forwards it to TVE that checks it (so that it doesn't obfuscate random / invalid data) and if it's valid, it encrypts it with the obfuscation key (currently compile-time mock, but planned to be periodically regenerated
+by another type of enclave) and returns the obfuscated transaction to TQE that forwards it to the client.
+
+In the first flow, TQE only talks to the TVE's application wrapper that handles the persistence -- it can unseal
+the transaction data, because the key policy is MRSIGNER. 
+
+In the second flow, TVE holds the obfuscation key inside the enclave memory, so the payload goes to TVE.
+Currently, TVE cannot check everything, e.g. staked state or if a transaction output was spent or not
+-- in the future, it may internally have app hash components and at least require some lightweight proofs
+for these things.
+
+For the first flow, it's unnecessary for TQE to talk to TVE. For the second flow, it'll be desirable
+to do a more complete verification (currently there are a few hacks and workarounds).
+
+## Decision
+This will be a bit big change, so it can be done in several steps:
+
+1. separate out the storage functionality from chain-abci into chain-storage crate
```

Comment:
Do you foresee any source of data races if we have a global storage process being accessed by multiple other processes (if both processes are reading and writing same things)?

Reply:
We can always serialize the requests in the storage process.

Reply:
> We can always serialize the requests in the storage process.

What do you mean by **serialize**? Can you elaborate?

Reply:
> > We can always serialize the requests in the storage process.
> 
> What do you mean by **serialize**? Can you elaborate?

With a mutex, like what abci-rs does with the mutex on `Application`?

Reply:
> Do you foresee any source of data races if we have a global storage process being accessed by multiple other processes (if both processes are reading and writing same things)?

No -- for the ABCI storage, it's the same as now... which relies on ABCI server implementation + needs some double-checking:
https://github.com/crypto-com/chain/blob/master/chain-abci/src/app/mod.rs#L406
https://github.com/crypto-com/chain/blob/master/chain-abci/src/app/mod.rs#L427

For the enclave storage, it'll be data in a different column, so no conflict with ABCI stuff... and it's a simple mapping: TXID -> sealed tx data payload 
added with deliver_tx (persisted in commit).
looked up in check_tx / deliver_tx.

Currently not deleted -- we may later implement some storage pruning (e.g. if the node isn't exposed for client querying, it could delete sealed transaction data if all outputs were spent)